### DISCUSSION TITLE
Straight through estimator (STE) for LDPC5GEncoder

### DISCUSSION THREAD

**Discussion by csankar69** [NVlabs/sionna#700] [General]
I am not sure why the following STE inside a class definition is not backproping gradients - I get a 'Gradients do not exist' warning. What is the correct way to pass gradients through an LDPC encoder?
```import tensorflow as tf
class model(Model):
  def __init__(..):
    ..

  @tf.custom_gradient
  def steLdpcEncoder(self, x):
    y = self.ldpcEncoder(tf.round(x))
    def grad(dy):
      return dy
    return y, grad
  
  def call(...):
    x = ...
    encodedBits = self.steLdpcEncoder(x)

**Comment by SebastianCa** [comment]
Hi @csankar69,
Without knowing the details of your experiment, it’s hard to give specific advice. Could you provide a minimal example that reproduces the issue? I quickly tested accessing the gradients, and it seems to work in my setup.

What’s the motivation for using an STE for the encoder? Conceptually, I’m not convinced this produces meaningful gradients (encoding is binary i.e. XOR operations). For instance, consider a toy example where the encoder only interleaves the input without adding redundancy. In such a case, your gradients would completely ignore this interleaving step.

I'll transfer this thread to discussions as it is not an issue with Sionna.

**Comment by csankar69** [comment]
Hi @SebastianCa,
Thanks for your response. I am trying to train a source encoder/decoder model that goes through the 5G transmission chain:
source encoder (Keras model) -> LDPC encoder -> modulation mapper -> resource grid mapper -> channel -> receiver -> modulation demapper -> LDPC decoder (hard_out=False) -> source decoder (Keras model) -> loss fn.
For the end-to-end training, I need to backprop the gradients through the chain and hence have to overcome the LDPC encoder issue. In my code above, x is the output of the source encoder. When I have the above steLdpcEncoder(), the gradients are not reaching the source encoder model (getting the 'Gradients do not exist' warning).

**Reply by SebastianCa** [reply]
Does it work to train the setup without the FEC? As said earlier, I'm not sure if your approach will lead to good results (as the encoder applies binary operations).

However, technically the following snippet seems to work 

```
batch_size = 1
k = 12
n = 24

class model(Model):
  def __init__(self,k, n):
    super().__init__()

    self.enc = LDPC5GEncoder(k,n)

  @tf.custom_gradient
  def steLdpcEncoder(self, x):
    y = self.enc(tf.round(x))
    def grad(dy):
      return dy
    return y, grad

  def call(self, x):
    c = self.steLdpcEncoder(x)
    return c

x = tf.Variable(tf.ones((batch_size, k)), dtype=tf.float32)

m = model(k,n)
with tf.GradientTape() as tape:
    y = m(x)
grads = tape.gradient(y, x)
print(grads)
```
Note that there will be also a dimension mismatch as the encoder maps k bits to n bits, however, your STE ignores this fact.

**Comment by csankar69** [comment]
@SebastianCa , yes I am aware that the encoder applies binary operations - how do I backprop the gradient past it (as in my case to the source encoder)?
There is a dimension mismatch and I need to modify `dy` to take care of this (as done in a simple manner in the code below).
The code below backprops the gradient through `steLdpcEncoder ` **when I don't have the modulation mapper/demapper in the chain** but when I have them - i.e., **uncomment the 2 lines** (`c = self.modMapper(c) and c = self.modDemapper([c, 1.0])`) it doesn't backprop (I don't see the statement "Came here" in `def grad(dy)` printed and also get a warning about gradients not existing for the the source encoder layers):
```py
import tensorflow as tf
from tensorflow.keras import Model, Sequential
from tensorflow.keras.layers import Layer, Dense
from tensorflow import abs, einsum, square, maximum, norm, float32
from sionna.fec.ldpc.encoding import LDPC5GEncoder
from sionna.fec.ldpc.decoding import LDPC5GDecoder
from sionna.mapping import Constellation, Mapper, Demapper

class sourceEncoder(Layer):
    def __init__(self, config):
        super(sourceEncoder, self).__init__()
        self.out = Sequential([Dense(config['innerDim'], activation='relu'),
                               Dense(config['srcCodeDim'], activation='sigmoid')])
    def call(self, x, training=True):
        return self.out(x)

class sourceDecoder(Layer):
    def __init__(self, config):
        super(sourceDecoder, self).__init__()
        self.out = Sequential([Dense(config['innerDim'], activation='relu'),
                               Dense(2*config['inputDim'], activation='relu')])
    def call(self, x, training=True):
        return self.out(x)

class e2eSystem(Model):
  def __init__(self, config):
    super().__init__()
    self.k = config['ldpc_k']
    self.coderate = 0.33
    self.n = int(self.k/self.coderate)
    self.sourceEncoder = sourceEncoder(config)
    self.enc = LDPC5GEncoder(self.k, self.n)
    self.constellation = Constellation("qam", config['numBitsPerSymbol'], trainable=True)
    self.modMapper = Mapper(constellation=self.constellation)
    self.modDemapper = Demapper("app", constellation=self.constellation)
    self.dec = LDPC5GDecoder(self.enc, hard_out=False)
    self.sourceDecoder = sourceDecoder(config)
    
  @tf.custom_gradient
  def steLdpcEncoder(self, x):
    y = self.enc(tf.round(x))
    def grad(dy):
      print("Came here")
      return dy[:, :self.k]
    return y, grad

  def call(self, x):
    x = tf.concat([tf.math.real(x), tf.math.imag(x)], axis=-1)
    c = self.sourceEncoder(x)
    c = self.steLdpcEncoder(c)
    #c = self.modMapper(c)
    #c = self.modDemapper([c, 1.0])
    c = self.dec(c)
    c = self.sourceDecoder(c)
    return c

def sgcs(V1, V2): #squared generalized cosine similarity
    absProduct = abs(einsum('ij,ij->i', tf.math.conj(V1), V2))
    return square(absProduct/maximum(1e-10, tf.cast(norm(V1, axis=1)*norm(V2, axis=1), dtype=float32))) #1e-10 to avoid div by 0

def trainModel(config):
    batchSize = 2
    optimizer = tf.keras.optimizers.Adam()
    model = e2eSystem(config)
    x = tf.complex(tf.random.normal([batchSize, config['inputDim']]),  tf.random.normal([batchSize, config['inputDim']]))
    
    with tf.GradientTape() as tape:
        y = model(x)
        y_real, y_imag = tf.split(y, 2, axis=-1)
        y = tf.complex(y_real, y_imag)
        loss = -sgcs(x, y)
    weights = model.trainable_weights
    gradients = tape.gradient(loss, weights)
    optimizer.apply_gradients(zip(gradients, weights))

if __name__ == "__main__":
    config = {'numBitsPerSymbol':4,
              'inputDim':32,
              'ldpc_k':32,
              'innerDim':32,
              'srcCodeDim':32}
    trainModel(config)
```
With the mod/demapper **excluded** (commented out):
```
for g in gradients:
    if g is not None:
        print(g.shape)
    else:
        `print(type(g))`
(64, 32)
(32,)
(32, 32)
(32,)
(32, 32)
(32,)
(32, 64)
(64,)
```
With the mod/demapper **included**:
```
for g in gradients:
    if g is not None:
        print(g.shape)
    else:
        print(type(g))
<class 'NoneType'>
<class 'NoneType'>
<class 'NoneType'>
<class 'NoneType'>
(2, 16)
(32, 32)
(32,)
(32, 64)
(64,)
```
As seen above the gradients are not computed in the back-pass beyond the trainable modulation mapper (shape (2,16)), whereas they are computed when the mapper is excluded.

**Reply by SebastianCa** [reply]
The issue is not with the STE but with the Mapper. It provides gradients with respect to the constellation points, not the (binary) input bits. It’s important to consider: what would be the interpretation or utility of gradients with respect to the binary input bits in this context?

**Comment by csankar69** [comment]
Since the mapper is essentially a lookup operation based on the input binary sequence, I don't need the gradient with respect to the binary input. Tensorflow is able to compute the gradients of the loss with respect to the constellation points $\theta_{QAM}$. Is there a way I can set the gradient of $\theta_{QAM}$ with respect to the binary input _b_ as equal to 1? This will enable the gradient to back propagate upstream.
